spring-integration-stream/src/main/java/org/springframework/integration/stream/CharacterStreamReadingMessageSource.java

```diff
@@ -57,20 +96,30 @@ else if (bufferSize > 0) {
 		else {
 			this.reader = new BufferedReader(reader);
 		}
+		this.blockAndCloseContext = blockAndCloseContext;
 	}
 
 
+	@Override
 	public String getComponentType() {
 		return "stream:stdin-channel-adapter(character)";
 	}
 
+	@Override
 	public Message<String> receive() {
 		try {
 			synchronized (this.monitor) {
-				if (!this.reader.ready()) {
+				if (!this.blockAndCloseContext && !this.reader.ready()) {
 					return null;
 				}
 				String line = this.reader.readLine();
+				if (line == null) {
+					ConfigurableApplicationContext context = (ConfigurableApplicationContext) getApplicationContext();
+					if (context != null) {
+						this.logger.info("EOF on stream; closing context");
+						context.close();
```

Comment:
I don't like this.

Maybe `EndOfStreamEvent` and then an application will be free to chose what do to with the fact. 
Close `ctx`, just stop adapters or switch to another stream.
Anything possible, but not just aggressive unconditional `ctx.close()` .

I won't mind if you have enough arguments to stay with this solution.
Thanks